### PR TITLE
Feature/asv 1245 home dialog for promotions

### PR DESCRIPTION
**Pull request by paquiquinho**
**What does this PR do?**

This PR has 3 aims:
- Adding the icon to the home action bar when promotions are available. If there are available promotions, the icon shows a ticker on top of it showing how many of them you can claim;
-  Adding the home dialog on first launch of the app with promotions available;
- Implementing the request to the get promotions API;

**Database changed?**

No

**Where should the reviewer start?**

- [ ] HomePresenter.java

**How should this be manually tested?**

Check out the flows described above and check if everything matches what's coming in charles and works correctly;

**What are the relevant tickets?**

  Tickets related to this pull-request: [ASV-1245](<https://aptoide.atlassian.net/browse/ASV-1245>)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Partners build
- [ ] Unit tests pass
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/res/layout/promotions_home_dialog.xml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            android:layout_width="78dp"
+            android:layout_marginEnd="16dp"
+            android:layout_marginRight="16dp"
+            android:text="@string/skip"
+            style="@style/Aptoide.Button.Flat"
+            />
+        <Button
+            android:id="@+id/navigate_button"
+            android:layout_width="116dp"
+            android:text="check it out"
```
Comment (jdandrade) [inline app/src/main/res/layout/promotions_home_dialog.xml]: hardcoded

File: dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsRequest.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      TokenInvalidator tokenInvalidator, SharedPreferences sharedPreferences) {
+    final BaseBody body = new BaseBody();
+    return new GetPromotionAppsRequest(body, bodyInterceptor, httpClient, converterFactory,
+        tokenInvalidator, sharedPreferences);
+  }
+
+  @Override
+  protected Observable<GetPromotionAppsResponse> loadDataFromNetwork(V7.Interfaces interfaces,
+      boolean bypassCache) {
+    return interfaces.getPromotionApps(10, body, true);
```
Comment (jdandrade) [inline dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsRequest.java]: is the limit supposed to be hardcoded?

File: dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsResponse.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
+package cm.aptoide.pt.dataprovider.ws.v7.promotions;
+
+import cm.aptoide.pt.dataprovider.model.v7.BaseV7Response;
+import cm.aptoide.pt.dataprovider.model.v7.listapp.App;
+import java.util.List;
+
+public class GetPromotionAppsResponse extends BaseV7Response {
+
+  private Datalist datalist;
```
Comment (jdandrade) [inline dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsResponse.java]: We already have a Datalist object for the response in the project. Can check BaseV7EndlessDataListResponse and examples that extend it.

File: app/src/main/java/cm/aptoide/pt/home/Home.java [hunk 4]
```diff
@@ -48,4 +57,32 @@ public Completable actionBundleImpression(ActionBundle bundle) {
...
+      for (PromotionApp app : apps) {
+        if (!app.isClaimed()) {
+          promotions++;
+          appcValue += app.getAppcValue();
+        }
+      }
+    }
+    if (promotionsPreferencesManager.shouldShowPromotionsDialog()) {
+      showDialog = true;
+      promotionsPreferencesManager.dontShowPromotionsDialog();
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/Home.java]: The default value for this is true so if there is nothing set in the preferences when the first get is done, true will be returned;

File: app/src/main/java/cm/aptoide/pt/home/Home.java [hunk 5]
```diff
@@ -48,4 +57,32 @@ public Completable actionBundleImpression(ActionBundle bundle) {
...
+    } else {
+      showDialog = false;
+    }
+    return new HomePromotionsWrapper(hasPromotions, promotions, appcValue, showDialog);
+  }
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/Home.java]: fixed

File: app/src/main/java/cm/aptoide/pt/home/HomeFragment.java [hunk 6]
```diff
@@ -339,6 +351,40 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
     return RxView.clicks(promotionsButton);
   }
 
+  @Override public void showPromotionsHomeDialog(HomePromotionsWrapper wrapper) {
+    if (wrapper.shouldShowDialog()) {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/HomeFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/promotions/PromotionsPresenter.java [hunk 7]
```diff
@@ -22,7 +23,8 @@ private void getPromotionApps() {
 
     view.getLifecycleEvent()
         .filter(lifecycleEvent -> lifecycleEvent.equals(View.LifecycleEvent.CREATE))
-        .flatMap(__ -> promotionsManager.getPromotionApps())
+        .flatMapSingle(__ -> promotionsManager.getPromotionApps())
+        .observeOn(AndroidSchedulers.mainThread())
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/promotions/PromotionsPresenter.java]: This is Filipe's code, as soon as we merge everything here changes. Only threw that in there so I could test my flows without crashes;

File: app/src/main/java/cm/aptoide/pt/home/HomeFragment.java [hunk 8]
```diff
@@ -339,6 +351,40 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+      description.setText(getContext().getString(R.string.holidayspromotion_message_popup,
+          String.valueOf(wrapper.getTotalAppcValue())));
+      cancel.setOnClickListener(click -> alertDialog.dismiss());
+      navigate.setOnClickListener(click -> {
+        promotionsDialogClickEvent.onNext(null);
+        alertDialog.dismiss();
+      });
+      alertDialog.show();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/HomeFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/home/HomeFragment.java [hunk 9]
```diff
@@ -339,6 +351,40 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+      Button cancel = dialogView.findViewById(R.id.cancel_button);
+      Button navigate = dialogView.findViewById(R.id.navigate_button);
+      TextView description = dialogView.findViewById(R.id.description);
+      description.setText(getContext().getString(R.string.holidayspromotion_message_popup,
+          String.valueOf(wrapper.getTotalAppcValue())));
+      cancel.setOnClickListener(click -> alertDialog.dismiss());
+      navigate.setOnClickListener(click -> {
+        promotionsDialogClickEvent.onNext(null);
+        alertDialog.dismiss();
+      });
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/HomeFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/home/HomeFragment.java [hunk 10]
```diff
@@ -339,6 +351,40 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+    promotionsButton.setVisibility(View.VISIBLE);
+    if (homeWrapper.getPromotions() > 0) {
+      if (homeWrapper.getPromotions() < 10) {
+        promotionsTicker.setText(Integer.toString(homeWrapper.getPromotions()));
+      } else {
+        promotionsTicker.setText("9+");
+      }
+      promotionsTicker.setVisibility(View.VISIBLE);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/home/HomeFragment.java]: fixed

File: app/src/main/java/cm/aptoide/pt/promotions/PromotionsPreferencesManager.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+  public PromotionsPreferencesManager(PreferencesPersister preferencesPersister) {
+    this.preferencesPersister = preferencesPersister;
+  }
+
+  public boolean shouldShowPromotionsDialog() {
+    return preferencesPersister.get(SHOW_PROMOTIONS_DIALOG, true);
+  }
+
+  public void dontShowPromotionsDialog() {
```
[the remark this answers is not in the document]
Comment (paquiquinho) [reply app/src/main/java/cm/aptoide/pt/promotions/PromotionsPreferencesManager.java]: fixed

File: app/src/main/res/layout/promotions_home_dialog.xml (hunk 1 above)
Comment (paquiquinho) [reply app/src/main/res/layout/promotions_home_dialog.xml]: fixed

File: dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsResponse.java (hunk 3 above)
Comment (paquiquinho) [reply dataprovider/src/main/java/cm/aptoide/pt/dataprovider/ws/v7/promotions/GetPromotionAppsResponse.java]: fixed

**Pull request merged**